src/constitutive/TACSSmearedCompositeShellConstitutive.cpp

```diff
@@ -0,0 +1,600 @@
+
+  TacsScalar *fvals = new TacsScalar[nvals];
+  evalPlyTopBottomFailure(strain, fvals);
+
+  TacsScalar **dfvals = new TacsScalar *[nvals];
+  for (int i = 0; i < num_plies; i++) {
+    TacsScalar lamStrain[3], phi[3];
+
+    // plate bottom stress
+    dfvals[2 * i + 0] = new TacsScalar[nvars];
+    getLaminaStrain(strain, tb, lamStrain);
+    TacsScalar fval_bot =
+        ply_props[i]->failureStrainSens(ply_angles[i], lamStrain, phi);
+    dfvals[2 * i + 0][0] = phi[0];
+    dfvals[2 * i + 0][1] = phi[1];
+    dfvals[2 * i + 0][2] = phi[2];
+    dfvals[2 * i + 0][3] = tb * phi[0];
+    dfvals[2 * i + 0][4] = tb * phi[1];
+    dfvals[2 * i + 0][5] = tb * phi[2];
+    dfvals[2 * i + 0][6] = dfvals[2 * i + 0][7] = dfvals[2 * i + 0][8] = 0.0;
+
+    // plate top stress
+    dfvals[2 * i + 1] = new TacsScalar[nvars];
+    getLaminaStrain(strain, tt, lamStrain);
+    TacsScalar fval_top =
+        ply_props[i]->failureStrainSens(ply_angles[i], lamStrain, phi);
+    dfvals[2 * i + 1][0] = phi[0];
+    dfvals[2 * i + 1][1] = phi[1];
+    dfvals[2 * i + 1][2] = phi[2];
+    dfvals[2 * i + 1][3] = tt * phi[0];
+    dfvals[2 * i + 1][4] = tt * phi[1];
+    dfvals[2 * i + 1][5] = tt * phi[2];
+    dfvals[2 * i + 1][6] = dfvals[2 * i + 1][7] = dfvals[2 * i + 1][8] = 0.0;
+  }
+
+  TacsScalar ks_val =
+      ksAggregationSensProduct(fvals, nvals, nvars, ks_weight, dfvals, sens);
+
+  for (int i = 0; i < nvals; i++) {
+    delete[] dfvals[i];
+  }
+
+  delete[] fvals;
+  delete[] dfvals;
+
+  return ks_val;
+}
+
+// Add the derivative of the failure criteria w.r.t. the design variables
+void TACSSmearedCompositeShellConstitutive::addFailureDVSens(
+    int elemIndex, TacsScalar scale, const double pt[], const TacsScalar X[],
+    const TacsScalar e[], int dvLen, TacsScalar dfdx[]) {
+  int index = 0;
+  if (thickness_dv_num >= 0 && dvLen >= 1) {
+    int nvals = 2 * num_plies;
+    // Compute the total thickness of the laminate
+    TacsScalar tb = -0.5 * thickness;
+    TacsScalar tt = 0.5 * thickness;
+
+    TacsScalar *fvals = new TacsScalar[nvals];
+    evalPlyTopBottomFailure(e, fvals);
+    TacsScalar *dks_vals = new TacsScalar[nvals];
+    TacsScalar ks_val = ksAggregationSens(fvals, nvals, ks_weight, dks_vals);
+
+    for (int i = 0; i < num_plies; i++) {
+      TacsScalar lamStrain[3], phi[3];
+
+      // plate bottom stress
+      getLaminaStrain(e, tb, lamStrain);
+      TacsScalar fval_bot =
+          ply_props[i]->failureStrainSens(ply_angles[i], lamStrain, phi);
+      dfdx[index] += -0.5 * scale * dks_vals[2 * i + 0] *
+                     (phi[0] * e[3] + phi[1] * e[4] + phi[2] * e[5]);
+
+      // plate top stress
+      getLaminaStrain(e, tt, lamStrain);
+      TacsScalar fval_top =
+          ply_props[i]->failureStrainSens(ply_angles[i], lamStrain, phi);
+      dfdx[index] += 0.5 * scale * dks_vals[2 * i + 1] *
+                     (phi[0] * e[3] + phi[1] * e[4] + phi[2] * e[5]);
+    }
+
+    delete[] fvals;
+    delete[] dks_vals;
```

Comment:
Same comment as above with regard to allocating and destroying memory in functions

Reply:
See above